### ISSUE TITLE
needless_lifetimes points at the wrong lifetime

### ISSUE THREAD

**Issue by audunska** [rust-lang/rust-clippy#10093]
### Summary

The `needless_lifetimes` lint suggests correctly to elide the `'name` lifetime. However, the ascii art points at the `'pool` lifetime, which cannot be elided.

### Reproducer

The following code is simplified from a `sqlx::query_as!` macro invocation. It could possibly be minified more by looking at the sqlx trait implementations. The sqlx version is 0.6.

```rust
use futures::Stream;
use sqlx::{postgres::PgArguments, Arguments, Postgres, Row};

pub fn bad_lint<'name, 'pool>(
    name: &'name str,
    pool: &'pool sqlx::PgPool,
) -> impl Stream<Item = Result<i64, sqlx::Error>> + 'pool {
    let mut query_args = PgArguments::default();
    query_args.add(&name);
    let q = sqlx::query_with::<Postgres, _>("SELECT number from table WHERE name = $1", query_args)
        .try_map(|row| row.try_get_unchecked::<i64, _>(0usize));
    q.fetch(pool)
}
```

Running clippy on 1.67.0-beta.2 correctly gives a warning that the lifetime `'name` can be elided, but the ascii art points at the other lifetime `'pool`:
```
warning: the following explicit lifetimes could be elided: 'name
 --> src/main.rs:4:1
  |
4 | / pub fn bad_lint<'name, 'pool>(
5 | |     name: &'name str,
6 | |     pool: &'pool sqlx::PgPool,
7 | | ) -> impl Stream<Item = Result<i64, sqlx::Error>> + 'pool {
  | |_________________________________________________________^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_lifetimes
  = note: `#[warn(clippy::needless_lifetimes)]` on by default
```

I would expect it to point at the `'name` lifetime.

### Version

```text
rustc 1.67.0-beta.2 (352eb59a4 2022-12-13)
binary: rustc
commit-hash: 352eb59a4c33abf739914422f2ad975925750146
commit-date: 2022-12-13
host: x86_64-unknown-linux-gnu
release: 1.67.0-beta.2
LLVM version: 15.0.6
```

**Comment by Niki4tap** [comment]
Thanks for the report!

Although this isn't *exactly* a bug, I see how this may cause confusion.

The issue is mostly just how clippy (rustc) highlights multiline spans, in your case, the function signature is spread across multiple lines, which causes rustc to highlight it not like it does in the most cases (with arrows `^^^^`), but from start to end (`/` denoting the start of the span, at `pub fn`, and the little arrow `^`, denoting the end of the span).

In your particular case, the end of the span just so happens to also be a lifetime, which is probably why you thought that this is a bug, but just to clarify: **the arrow is pointing at the end of the function signature, not the lifetime specifically**.

This also depends on the formatting, because if you collapse the function signature into one line you will see that the whole function signature is highlighted, not particular lifetimes.

```
error: the following explicit lifetimes could be elided: 'needless
  --> $DIR/right_span.rs:1:1
   |
LL | fn issue10093<'needless, 'needed>(x: &'needless str, y: &'needed str) -> impl ::core::fmt::Display + 'needed {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

This will become even more clear if you reorder the bounds:

```
error: the following explicit lifetimes could be elided: 'needless
  --> $DIR/right_span.rs:1:1
   |
LL | / fn issue10093<'needless, 'needed>(
LL | |     x: &'needless str,
LL | |     y: &'needed str
LL | | ) -> impl 'needed + ::core::fmt::Display {
   | |________________________________________^
```

This probably needs a decision of someone on the clippy team, on how exactly to approach this issue. Ideally, clippy would point out exactly which lifetimes can be elided and note the individual lifetimes that need to be removed in a suggestion, but I imagine that would be pretty hard to implement, and then it's more of an enhancement request than a bug report :)

I'm not trying to say that you should close the issue, just documenting the potential implementation a bit for anyone who will try to approach this issue in the future.

With that said, @rustbot label +S-needs-discussion +C-question +A-ui +E-help-wanted

**Comment by audunska** [comment]
Thanks for the explanation! I think I've always read the little arrow as pointing at a specific part of the span, not the entire line. So this is indeed a bit confusing. Although I see now how the  `|` on the left-hand side along with the `_` sort of tries to paint a box around the function signature.

One slight improvement might be to move the little arrow one space to the right, to point at the whitespace between the signature and the `{`? Then it would look more like we are boxing in the signature. Although I understand we can't depend on there being whitespace there.
